src/modm/platform/dac/stm32/dac.hpp.in

```diff
@@ -0,0 +1,224 @@
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_DAC{{ id }}_HPP
+#define MODM_STM32_DAC{{ id }}_HPP
+
+#include <cstdint>
+#include "../device.hpp"
+#include <modm/platform/gpio/connector.hpp>
+#include <modm/platform/clock/rcc.hpp>
+#include <modm/src/modm/math/units.hpp>
+
+namespace modm::platform
+{
+
+/**
+ * Digital-to-Analog Converter (DAC{{ id }})
+ *
+ * Supports simple synchronous access to the 2-channel DAC found on most STM32
+ * microcontrollers.
+ *
+ * @author Jeff McBride
+ * @ingroup modm_platform_dac{% if id %} modm_platform_dac_{{id}}{% endif %}
+ */
+class Dac{{ id }} {
+public:
+	/**
+%% if dual_channel
+	 * The DAC has two output channels, Channel1 and Channel2
+%% else
+	 * The DAC has one output channel, Channel1
+%% endif
+	 */
+	enum class Channel : uint8_t
+	{
+		Channel1 = 0,
+%% if dual_channel
+		Channel2 = 1
+%% endif
+	};
+
+%% if has_mcr
+	enum class Mode
+	{
+		ExternalWithBuffer = 0,
+		ExternalInternalWithBuffer = DAC_MCR_MODE1_0,
+		ExternalWithoutBuffer = DAC_MCR_MODE1_1,
+		Internal = DAC_MCR_MODE1_0 | DAC_MCR_MODE1_1
+	};
+%% endif
+
+	template< template<Peripheral _> class... Signals >
+	static void
+	connect()
+	{
+		using Connector = GpioConnector<Peripheral::Dac{{ id }}, Signals...>;
+		Connector::connect();
+	}
+
+%% if target.family == "g4"
+	/**
+	 * Initialize the D/A converter.
+	 *
+	 * Enables the RCC clock output for the DAC
+	 *
+	 * @tparam SystemClock System clock struct
+	 */
+	template<typename SystemClock>
+	static inline void initialize();
+
+%% else
+	/**
+	 * Initialize the D/A converter.
+	 *
+	 * Enables the RCC clock output for the DAC
+	 */
+	static void initialize();
+
+%% endif
+
+	/** Enable DAC output for the given channel
+	 *
+	 * @param chan    The channel to be enabled
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void enableChannel(Channel chan);
+
+	/**
+	 * Disable DAC output for the given channel
+	 *
+	 * @param chan    The channel to be disabled
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void disableChannel(Channel chan);
+
+%% if has_mcr
+	/**
+	 * Control the output mode
+	 *
+	 * @param chan  The channel to configure the mode for
+	 * @param mode  The channel output mode
+	 *
+	 * @warning The DAC channel will be disabled when setting the mode
+	 */
+	static void setMode(Channel chan, Mode mode);
+
+%% else
+	/**
+	 * Control the output buffer setting
+	 *
+	 * @param chan    The channel to enable/disable the output buffer on
+	 * @param enable  true to turn on the output buffer, false to turn it off
+	 *
+	 * @pre The DAC clock must be enabled with initialize()
+	 */
+	static void enableOutputBuffer(Channel chan, bool enable);
+
+%% endif
+
+	/**
+	 * Set the output voltage for a DAC channel
+	 *
+	 * @param chan    The channel to set
+	 * @param value   The 12-bit DAC output value, range 0-4095.
```

Comment:
I would prefer to use the full range of uint16_t as input (and the left-aligned data register `DHR12L1`) to be able to use the same interface for different DACs in the future and better portability.

In any case, it could be useful to put the precision of the DAC somewhere in the code:
`static constexpr uint8_t PrecisionBit = 12;`

Reply:
> I would prefer to use the full range of uint16_t as input (and the left-aligned data register DHR12L1) to be able to use the same interface for different DACs in the future and better portability.

It would be quite evil to change the data format to left-justified for an already existing driver. The benefit seems rather small and the choice seems a bit arbitrary as well. Let's say we have an 8 bit or 18 bit DAC, in both cases 16 bit right aligned data is not the most useful choice. It would be fine for me to state in some `constexpr` what the data format is and the user code can deal with it.

> In any case, it could be useful to put the precision of the DAC somewhere in the code:
static constexpr uint8_t PrecisionBit = 12;

:+1:

Reply:
Ok 👍🏽